### ISSUE TITLE
Unable to convert MobileNetV3Large to TFLite with integer quantization

### ISSUE THREAD

**Issue by Musyyab** [tensorflow/tensorflow#83798]
**1. System information**

1.  Operating System and Distribution: Windows 10 and Windows 10 WSL

1. TensorFlow Installation Method: pip package (version 2.10 on Windows 10) and version 2.16.2 on WSL

**2. Code**

import tensorflow as tf
import numpy as np
from tensorflow.keras.applications import MobileNetV3Large
from tensorflow.keras.applications.mobilenet_v3 import preprocess_input
import matplotlib.pyplot as plt
from scipy.stats import pearsonr

# Generate a sample image for testing
test_image = np.random.normal(loc=127.5, scale=50, size=(1, 224, 224, 3))
test_image = np.clip(test_image, 0, 255).astype(np.float32)
preprocessed_image = preprocess_input(test_image.copy())

# Load the MobileNetV3Large model with ImageNet weights
model = MobileNetV3Large(
    weights='imagenet',
    include_top=True,
    input_shape=(224, 224, 3)
)

# Get predictions from the original model
original_pred = model.predict(preprocessed_image, verbose=0)

# Convert the model to TFLite with optimizations
converter = tf.lite.TFLiteConverter.from_keras_model(model)
converter.optimizations = [tf.lite.Optimize.DEFAULT]  # Enable dynamic range quantization
converter.target_spec.supported_ops = [
    tf.lite.OpsSet.TFLITE_BUILTINS,
    tf.lite.OpsSet.SELECT_TF_OPS
]
converter.experimental_new_converter = True

# Convert the model
tflite_model = converter.convert()

# Load the TFLite model and allocate tensors
interpreter = tf.lite.Interpreter(model_content=tflite_model)
interpreter.allocate_tensors()

# Get input and output details
input_details = interpreter.get_input_details()
output_details = interpreter.get_output_details()

# Run inference with the TFLite model
interpreter.set_tensor(input_details[0]['index'], preprocessed_image)
interpreter.invoke()
tflite_pred = interpreter.get_tensor(output_details[0]['index'])

# Calculate correlation between original and TFLite predictions
correlation, _ = pearsonr(original_pred.flatten(), tflite_pred.flatten())

# Visualize results
plt.figure(figsize=(12, 6))

# Scatter plot: Original vs Quantized Predictions
plt.subplot(1, 2, 1)
plt.scatter(original_pred.flatten(), tflite_pred.flatten(), alpha=0.5)
plt.plot([original_pred.min(), original_pred.max()],
         [original_pred.min(), original_pred.max()],
         'r--', label=f'Perfect Correlation\nActual: {correlation:.4f}')
plt.title('Original vs Quantized Predictions')
plt.xlabel('Original Model Predictions')
plt.ylabel('Quantized Model Predictions')
plt.legend()

# Distribution of prediction differences
plt.subplot(1, 2, 2)
plt.hist(np.abs(original_pred.flatten() - tflite_pred.flatten()),
         bins=50, alpha=0.75, label='Prediction Differences')
plt.title('Distribution of Prediction Differences')
plt.xlabel('|Original - Quantized|')
plt.ylabel('Frequency')
plt.legend()

plt.tight_layout()
plt.show()

# Print results
print(f"\nResults:")
print(f"Prediction Correlation: {correlation:.4f}")
print(f"Original Model Size: {len(model.get_weights()) / 1024 / 1024:.2f} MB")
print(f"Quantized Model Size: {len(tflite_model) / 1024 / 1024:.2f} MB")
print(f"Size Reduction: {(1 - len(tflite_model) / len(model.get_weights())) * 100:.1f}%")

**Failure after conversion**

TF 2.10 in Win10 Log:
Model produces wrong results. See plot made from code:
image
![220px-3-bit_resolution_analog_comparison](https://github.com/user-attachments/assets/892a3b56-9f38-4e64-9464-4506db55bcfb)

TF2.16 in WSL:
Model fails to convert. Gets error: LLVM ERROR: Failed to infer result type(s). (see log)

**5. (optional) Any other info / logs**

Updated Code

python
Copy code
import tensorflow as tf
import numpy as np
from tensorflow.keras.applications import MobileNetV3Large
from tensorflow.keras.applications.mobilenet_v3 import preprocess_input
import matplotlib.pyplot as plt
from scipy.stats import pearsonr

# Check TensorFlow version
print(f"TensorFlow version: {tf.__version__}")
print(f"Python version: {sys.version}\n")

# Generate a sample image for testing
test_image = np.random.normal(loc=127.5, scale=50, size=(1, 224, 224, 3))
test_image = np.clip(test_image, 0, 255).astype(np.float32)
preprocessed_image = preprocess_input(test_image.copy())

# Load the MobileNetV3Large model
model = MobileNetV3Large(weights='imagenet', include_top=True, input_shape=(224, 224, 3))

# Get original model predictions
original_pred = model.predict(preprocessed_image, verbose=0)

# Convert to TFLite
converter = tf.lite.TFLiteConverter.from_keras_model(model)
converter.optimizations = [tf.lite.Optimize.DEFAULT]  # Enable dynamic range quantization
converter.target_spec.supported_ops = [
    tf.lite.OpsSet.TFLITE_BUILTINS,
    tf.lite.OpsSet.SELECT_TF_OPS  # Allow TensorFlow ops not natively supported in TFLite
]

try:
    tflite_model = converter.convert()
except Exception as e:
    print(f"Error during TFLite conversion: {e}")
    exit()

# Load TFLite model into an interpreter
interpreter = tf.lite.Interpreter(model_content=tflite_model)
interpreter.allocate_tensors()

# Set up input and output details
input_details = interpreter.get_input_details()
output_details = interpreter.get_output_details()

# Ensure data type compatibility with TFLite
input_data = np.array(preprocessed_image, dtype=input_details[0]['dtype'])
interpreter.set_tensor(input_details[0]['index'], input_data)

# Invoke TFLite interpreter
interpreter.invoke()
tflite_pred = interpreter.get_tensor(output_details[0]['index'])

# Calculate correlation
correlation, _ = pearsonr(original_pred.flatten(), tflite_pred.flatten())

# Visualize results
plt.figure(figsize=(10, 5))

# Scatter plot: Original vs Quantized predictions
plt.subplot(1, 2, 1)
plt.scatter(original_pred.flatten(), tflite_pred.flatten(), alpha=0.5)
plt.plot([original_pred.min(), original_pred.max()],
         [original_pred.min(), original_pred.max()],
         'r--', label=f'Perfect Correlation\nActual: {correlation:.4f}')
plt.title('Original vs Quantized Predictions')
plt.xlabel('Original Model')
plt.ylabel('Quantized Model')
plt.legend()

# Distribution of differences
plt.subplot(1, 2, 2)
plt.hist(np.abs(original_pred.flatten() - tflite_pred.flatten()), bins=50, alpha=0.75, label='Prediction Differences')
plt.title('Distribution of Prediction Differences')
plt.xlabel('|Original - Quantized|')
plt.ylabel('Count')
plt.legend()

plt.tight_layout()
plt.show()

# Print results
print(f"Prediction correlation: {correlation:.4f}")
print(f"Original model size (MB): {sum([w.size * w.itemsize for w in model.get_weights()]) / (1024 * 1024):.2f}")
print(f"TFLite model size (MB): {len(tflite_model) / (1024 * 1024):.2f}")
Key Updates:
Version Compatibility: Checked TensorFlow version to avoid using deprecated flags or unsupported options.
Error Handling: Added a try-except block for converter.convert() to catch issues during TFLite conversion.
Data Type Matching: Explicitly cast preprocessed_image to match input_details[0]['dtype'], addressing potential dtype mismatches.
Model Size Calculation: Used a consistent method to calculate and print the original and TFLite model sizes.
Robustness: Incorporated better diagnostics for potential errors.
**How to Run:**
Ensure that your environments (Windows and WSL) have the appropriate TensorFlow version and GPU drivers (if applicable).
For WSL, ensure cuda, cudnn, and TensorRT are properly installed or switch to CPU mode if GPU support is unavailable.
Execute the code with the updated version in both environments and observe the logs for differences or issues.

**Comment by gaikwadrahul8** [comment]
Hi, @Musyyab 
I apologize for the delayed response, thank you for bringing this issue to our attention, I tried downgrading the TensorFlow version to `2.14.1` and it seems like code is working as expected please refer this similar issue https://github.com/tensorflow/tensorflow/issues/83754 so could you please give it try with TensorFlow version 2.14.1 by installing using this command `pip install tensorflow==2.14.1 `for Windows 10 WSL and let us know is it working as expected or not ?

If issue still persists or I have missed something here please let me know with updated error log for further investigation.

Thank you for your cooperation and understanding.
